src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs

```diff
@@ -18,38 +19,89 @@ namespace Microsoft.Azure.WebJobs.Script.Workers
 {
     internal static class ScriptInvocationContextExtensions
     {
-        public static InvocationRequest ToRpcInvocationRequest(this ScriptInvocationContext context, bool isTriggerMetadataPopulatedByWorker, ILogger logger, Capabilities capabilities)
+        public static async Task<InvocationRequest> ToRpcInvocationRequest(this ScriptInvocationContext context, bool isTriggerMetadataPopulatedByWorker, ILogger logger, Capabilities capabilities)
         {
-            InvocationRequest invocationRequest = new InvocationRequest()
+            var invocationRequest = new InvocationRequest
             {
                 FunctionId = context.FunctionMetadata.FunctionId,
                 InvocationId = context.ExecutionContext.InvocationId.ToString(),
                 TraceContext = GetRpcTraceContext(context.Traceparent, context.Tracestate, context.Attributes, logger),
             };
 
-            foreach (var pair in context.BindingData)
+            var rpcValueCache = new Dictionary<object, TypedData>();
+
+            foreach (var input in context.Inputs)
             {
-                if (pair.Value != null)
+                if (!rpcValueCache.TryGetValue(input.val, out TypedData rpcValue))
                 {
-                    if ((pair.Value is HttpRequest) && isTriggerMetadataPopulatedByWorker)
-                    {
-                        continue;
-                    }
-                    invocationRequest.TriggerMetadata.Add(pair.Key, pair.Value.ToRpc(logger, capabilities));
+                    rpcValue = await input.val.ToRpc(logger, capabilities);
                 }
+
+                var parameterBinding = new ParameterBinding
+                {
+                    Name = input.name,
+                    Data = rpcValue
+                };
+                invocationRequest.InputData.Add(parameterBinding);
             }
-            foreach (var input in context.Inputs)
+
+            foreach (var pair in context.BindingData)
             {
-                invocationRequest.InputData.Add(new ParameterBinding()
+                if (ShouldSkipBindingData(pair, context, isTriggerMetadataPopulatedByWorker))
                 {
-                    Name = input.name,
-                    Data = input.val.ToRpc(logger, capabilities)
-                });
+                    continue;
+                }
+
+                if (!rpcValueCache.TryGetValue(pair.Value, out TypedData rpcValue))
+                {
+                    rpcValue = await pair.Value.ToRpc(logger, capabilities);
+                    rpcValueCache.Add(pair.Value, rpcValue);
+                }
+
+                invocationRequest.TriggerMetadata.Add(pair.Key, rpcValue);
             }
 
             return invocationRequest;
         }
 
+        private static bool ShouldSkipBindingData(KeyValuePair<string, object> bindindData, ScriptInvocationContext context, bool isTriggerMetadataPopulatedByWorker)
```

Comment:
The amount of arguably unnecessary binding data we were sending was a big perf drain. E.g. we currently duplicate the request body 3 times in the request payload (twice in binding data, and once in inputs). In addition, the headers/query are also separate binding data members.

@pragnagopa I believe what I've done here is safe, but I'd like your input. On the worker side, binding data is only used for resolving parameters, right?

Reply:
@yojagad did work to remove this under a capability flag (RpcHttpBodyOnly) and I think python was the reason it was under a capability flag? Yogesh do you remember?

Reply:
Unfortunately this was an oversight in the initial Rpc design. Yes what @mhoeger pointed out is right.  We can use "RpcHttpBodyOnly" to skip sending duplicate data. @yojagad has more context on this.

Reply:
yeah, it's under RpcHttpBodyOnly (In RpcMessageConversionExtensions). If a language worker indicates it's capable of constructing raw body from the body provided, it sends only body. 

---
In reply to: [414888273](https://github.com/Azure/azure-functions-host/pull/5897#discussion_r414888273) [](ancestors = 414888273)

Reply:
For the Java worker, neither RpcHttpBodyOnly nor RpcHttpTriggerMetadataRemoved are set, meaning we don't get those optimizations. Can we update the Java worker to support these? However, for the latter, as you can see in my changes, I'm removing all the duplicate HttpRequest data from BindingData, since we're already sending via Inputs.

Reply:
Pinged @amamounelsayed for java worker updates on issue: https://github.com/Azure/azure-functions-java-worker/issues/319 .

Reply:
Thank you @pragnagopa  and @mathewc  we ware still investigating if we can add it to this sprint or not. It is on top of our list. Will get back once we know more info. Thank you!

Reply:
Ahmed released 1.6.1 of the Java worker with the required changes. I've pulled that in.